python/pyspark/sql/session.py

```diff
@@ -681,12 +694,14 @@ def createDataFrame(self, data, schema=None, samplingRatio=None, verifySchema=Tr
 
     def _create_dataframe(self, data, schema, samplingRatio, verifySchema):
         if isinstance(schema, StructType):
+            no_need_to_prepare = not verifySchema
             verify_func = _make_type_verifier(schema) if verifySchema else lambda _: True
 
             def prepare(obj):
                 verify_func(obj)
                 return obj
```

Comment:
Hmm, doesn't `prepare` already verify the schema? Why we need to verify it again in converter?

Reply:
The prepare phase verification only covers when schema `isinstance` of StructType and DataType.

The conversion phase covers when schema is None and is inferred during the conversion phase.